pkg/blocks/fetchuri.go

```diff
@@ -0,0 +1,309 @@
+func (f *FetchURIStep) Execute(execCtx TTPExecutionContext) (*ExecutionResult, error) {
+	logging.L().Info("========= Executing ==========")
+
+	if err := f.fetchURI(execCtx); err != nil {
+		logging.L().Error(zap.Error(err))
+		return nil, err
+	}
+
+	logging.L().Info("========= Result ==========")
+
+	return &ExecutionResult{}, nil
+}
+
+// fetchURI executes the FetchURIStep with the specified Location, Uri, and additional arguments,
+// and an error if any errors occur.
+func (f *FetchURIStep) fetchURI(execCtx TTPExecutionContext) error {
+	appFs := f.FileSystem
+	absLocal := f.Location
+
+	if appFs == nil {
+		var err error
+		appFs = afero.NewOsFs()
+		absLocal, err = FetchAbs(f.Location, f.WorkDir)
+		if err != nil {
+			return err
+		}
+	}
+
+	if ok, _ := afero.Exists(appFs, absLocal); ok && !f.Overwrite {
+		logging.L().Errorw("location exists, remove and retry", "location", absLocal)
+		return fmt.Errorf("location [%s] exists and overwrite is set to false. remove and retry", f.Location)
+	}
+
+	client := http.DefaultClient
+	if f.Proxy != "" {
+		proxyURI, err := url.Parse(f.Proxy)
+		if err != nil {
+			return err
+		} else if proxyURI.Host == "" || proxyURI.Scheme == "" {
+			return fmt.Errorf("invalid URI given for Proxy: %s", f.Proxy)
+		}
+		tr := &http.Transport{
+			Proxy: http.ProxyURL(proxyURI),
+		}
+		client = &http.Client{Transport: tr}
+	}
+
+	resp, err := client.Get(f.FetchURI)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	fHandle, err := appFs.Create(absLocal)
+	if err != nil {
+		return err
+	}
+	defer fHandle.Close()
+
+	_, err = io.Copy(fHandle, resp.Body)
+	if err != nil {
+		return err
+	}
+
+	logging.L().Debugw("wrote contents of URI to specified location", "location", absLocal, "uri", f.FetchURI)
+
+	return nil
+}
+
+// Validate validates the FetchURIStep. It checks that the
+// Act field is valid, Location is set with
+// a valid file path, and Uri is set.
+//
+// If Location is set, it ensures that the path exists and retrieves
+// its absolute path.
+//
+// **Returns:**
+//
+// error: An error if any validation checks fail.
+func (f *FetchURIStep) Validate(execCtx TTPExecutionContext) error {
+	if err := f.Act.Validate(); err != nil {
+		logging.L().Error(zap.Error(err))
+		return err
+	}
+
+	if f.FetchURI == "" {
+		err := errors.New("require FetchURI to be set with fetchURI")
+		logging.L().Error(zap.Error(err))
+		return err
+	}
+
+	if f.Location == "" {
+		err := errors.New("require Location to be set with fetchURI")
+		logging.L().Error(zap.Error(err))
+		return err
+	}
+
+	if f.Proxy != "" {
+		uri, err := url.Parse(f.Proxy)
+		if err != nil {
+			return err
+		} else if uri.Host == "" || uri.Scheme == "" {
+			return fmt.Errorf("invalid URI given for Proxy: %s", f.Proxy)
+		}
+	}
+
+	// Retrieve the absolute path to the file.
+	absLocal, err := FetchAbs(f.Location, f.WorkDir)
+	if err != nil {
+		logging.L().Error(zap.Error(err))
+		return err
+	}
+
+	_, err = os.Stat(absLocal)
```

Comment:
May want to use `f.FileSystem` here as well for consistency? If `FileSystem` is an afero in-memory fs in a test, this might fail